src/compiler/factory/nodeFactory.ts

```diff
@@ -4972,6 +4972,11 @@ namespace ts {
             node.transformFlags |=
                 propagateChildrenFlags(node.statements) |
                 propagateChildFlags(node.endOfFileToken);
+            node.referencedFiles = [];

```

Comment:
why are these new allocations instead of using emptyArray or emptyMap

Reply:
Because this is our node factory, which is publicly exposed (which is the point, I want these set on all factory-made nodes), and I _really_ don't trust our API consumers to not pollute those with what I've seen out in the wild. 😄

Reply:
but they are suppose to be readonly array typed fields...

Reply:
They're definitely supposed to be, but it's really common to make a `SourceFile` node and then mutate it (since it's the root node) to set the many fields not set by the constructor (even though the types don't strictly allow that).

Reply:
Wouldnt that be user issue then. Its not like we never use emptyArray for anything exposed by Public API

Reply:
I mean, sure, but it's not like we have had documentation that "fields x, y, and z are initialized by the factory to empty singletons which shouldn't be mutated" - and since the factory _didn't_ make these fields before, if someone had something to handle that in place, like 
```ts
const file = (input || factory.createSourceFile()) as Mutable<SourceFile>;
file.referencedFiles ||= [];
file.referencedFiles.push(other);
```
this would break them where they used to work if I reused the singletons here.